hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+		LOG.debug(OBJECT_STREAM.getMarker(),
+				"RecordStreamFileWriter::setStartWriteAtCompleteWindow: {}", startWriteAtCompleteWindow);
+	}
+
+	public boolean getStartWriteAtCompleteWindow() {
+		return this.startWriteAtCompleteWindow;
+	}
+
+	/**
+	 * Helper method that serializes a RecordStreamFile.
+	 * Uses deterministic serialization to ensure multiple invocations on the same
+	 * object lead to identical serialization.
+	 * @param recordStreamFileProtoBuilder
+	 * 		the RecordStreamFile object that needs to be serialized
+	 * @return
+	 * 		the serialized bytes
+	 */
+	private byte[] serialize(final RecordStreamFile.Builder recordStreamFileProtoBuilder) throws IOException {
+		final var recordStreamFileProto = recordStreamFileProtoBuilder.build();
+		final var result = new byte[recordStreamFileProto.getSerializedSize()];
+		final var output = CodedOutputStream.newInstance(result);
+		output.useDeterministicSerialization();
+		recordStreamFileProto.writeTo(output);
+		output.checkNoSpaceLeft();
+		return result;
+	}
+
+	private HashObject toProto(final Hash hash) {
+		return HashObject.newBuilder()
+				.setAlgorithm(HashAlgorithm.SHA_384)
+				.setLength(hash.getDigestType().digestLength())
+				.setHash(ByteString.copyFrom(hash.getValue()))
+				.build();
+	}
+
+	private void createSignatureFile(final File relatedRecordStreamFile) {
+		// get entire hash of current record stream file
+		final var entireHash = new Hash(streamDigest.digest(), DigestType.SHA_384);
+		// get metaData hash of current record stream file
+		final var metaHash = new Hash(metadataStreamDigest.digest(), DigestType.SHA_384);
+
+		// create proto messages for signature file
+		final var fileSignature = generateSignatureObject(entireHash);
+		final var metadataSignature = generateSignatureObject(metaHash);
+		final var signatureFile = SignatureFile.newBuilder()
+				.setFileSignature(fileSignature)
+				.setMetadataSignature(metadataSignature);
+
+		// create signature file
+		final var sigFilePath = generateSigFilePath(relatedRecordStreamFile);
+		try (final var fos = new FileOutputStream(sigFilePath)) {
+			fos.write(recordFileVersion); // version in signature files is 1 byte, compared to 4 in record files
+			signatureFile.build().writeTo(fos);
+			LOG.debug(OBJECT_STREAM_FILE.getMarker(),
+					"closeCurrentAndSign :: signature file saved: {}", sigFilePath);
+		} catch (IOException e) {
+			LOG.error(EXCEPTION.getMarker(),
+					"closeCurrentAndSign ::  :: Fail to generate signature file for {}",
+					fileNameShort, e);
+		}
+	}
+
+	private SignatureObject generateSignatureObject(final Hash hash) {
+		final var signature = signer.sign(hash.getValue());
+		return SignatureObject.newBuilder()
+				.setType(SignatureType.SHA_384_WITH_RSA)
+				.setLength(signature.length)
+				.setChecksum(101 - signature.length) // simple checksum to detect if at wrong place in the stream
+				.setSignature(ByteString.copyFrom(signature))
```

Comment:
There's a util method somewhere in the codebase Michael added that you can use to avoid an array copy here and elsewhere in the file.

Reply:
[Fixed](https://github.com/hashgraph/hedera-services/pull/3461/commits/4cc3f212e7d35ef866331a8df180ba914b6d3ca7).